<a href="https://colab.research.google.com/github/stephenebert/Springboard/blob/main/Mini_Project_Exploratory_Data_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# [NYC Taxi Data Analysis](https://github.com/springboard-curriculum/mec2-projects/blob/main/Student_MLE_MiniProject_EDA.ipynb)

Import relevant packages



In [61]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

**Part 1**

**1. Load the yellow_tripdata_2022-01.parquet file into Pandas**

In [62]:
df = pd.read_parquet('yellow_tripdata_2022-01.parquet')

**2. Print the first 5 rows of data**

In [81]:
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1         1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2         2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3         2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4         2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.80         1.0                  N           142           236   
1           2.10         1.0                  N           236            42   
2           0.97         1.0                  N           166           166   
3           1.09         1.0                  N           114            68   
4           4.30         1.0                  N            68           163   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0             1         14.5    3.0      0.5        3.65           0.0   
1             1          8.0    0.5      0.5        4.00           0.0   
2             1          7.5    0.5      0.5        1.76           0.0   
3             2          8.0    0.5      0.5        0.00           0.0   
4             1         23.5    0.5      0.5        3.00           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  airport_fee  
0                    0.3         21.95                   2.5          0.0  
1                    0.3         13.30                   0.0          0.0  
2                    0.3         10.56                   0.0          0.0  
3                    0.3         11.80                   2.5          0.0  
4                    0.3         30.30                   2.5          0.0

**3. How many rows are in the dataset? How many unique columns are in the dataset?**

In [64]:
df.shape

(2463931, 19)

There are **2463931 rows** and **19 columns**. Here we used the .shape to get this attribute (rows, columns), but there could be duplicates

We see it doesn't matter for this example, but, to be extra safe, we will use the .unique() method to arrive at the same answer

In [65]:
len(df.columns.unique()) #First we extract the columns, use the .unique() method to filter duplicates, then print the number of unique columns

19

In [66]:
len(df) #rows

2463931

**4. Which columns have NULL values and how many NULL values are present in each of these columns?**

In [70]:
df.isnull().sum()

VendorID                     0
tpep_pickup_datetime         0
tpep_dropoff_datetime        0
passenger_count          71503
trip_distance                0
RatecodeID               71503
store_and_fwd_flag       71503
PULocationID                 0
DOLocationID                 0
payment_type                 0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
improvement_surcharge        0
total_amount                 0
congestion_surcharge     71503
airport_fee              71503
dtype: int64

There are five columns that have null values are "**passenger_count**", "**RatecodeID**", "**store_and_fwd_flag**", "**congestion_surcharge**", and "**airport_fee**". There are **71503** null values present for each of these five columns.

The other fourteen columns have no null values as you can see

**5. Generate summary statistics using Pandas' describe method. Do you notice anything unusual in the dataset? Find at least one anomoly and try to come up with a hypothesis to explain it.**

In [68]:
df.describe()

VendorID        tpep_pickup_datetime       tpep_dropoff_datetime  \
count  2.463931e+06                     2463931                     2463931   
mean   1.707819e+00  2022-01-17 01:19:51.689724  2022-01-17 01:34:04.421902   
min    1.000000e+00         2008-12-31 22:23:09         2008-12-31 23:06:56   
25%    1.000000e+00         2022-01-09 15:37:41  2022-01-09 15:50:50.500000   
50%    2.000000e+00         2022-01-17 12:11:45         2022-01-17 12:23:49   
75%    2.000000e+00  2022-01-24 13:49:37.500000         2022-01-24 14:02:51   
max    6.000000e+00         2022-05-18 20:41:57         2022-05-18 20:47:45   
std    5.021375e-01                         NaN                         NaN   

       passenger_count  trip_distance    RatecodeID  PULocationID  \
count     2.392428e+06   2.463931e+06  2.392428e+06  2.463931e+06   
mean      1.389453e+00   5.372751e+00  1.415507e+00  1.660768e+02   
min       0.000000e+00   0.000000e+00  1.000000e+00  1.000000e+00   
25%       1.000000e+00   1.040000e+00  1.000000e+00  1.320000e+02   
50%       1.000000e+00   1.740000e+00  1.000000e+00  1.620000e+02   
75%       1.000000e+00   3.130000e+00  1.000000e+00  2.340000e+02   
max       9.000000e+00   3.061593e+05  9.900000e+01  2.650000e+02   
std       9.829686e-01   5.478714e+02  5.917573e+00  6.546806e+01   

       DOLocationID  payment_type   fare_amount         extra       mta_tax  \
count  2.463931e+06  2.463931e+06  2.463931e+06  2.463931e+06  2.463931e+06   
mean   1.635807e+02  1.194449e+00  1.294648e+01  1.006710e+00  4.914539e-01   
min    1.000000e+00  0.000000e+00 -4.800000e+02 -4.500000e+00 -5.000000e-01   
25%    1.130000e+02  1.000000e+00  6.500000e+00  0.000000e+00  5.000000e-01   
50%    1.620000e+02  1.000000e+00  9.000000e+00  5.000000e-01  5.000000e-01   
75%    2.360000e+02  1.000000e+00  1.400000e+01  2.500000e+00  5.000000e-01   
max    2.650000e+02  5.000000e+00  4.010923e+05  3.350000e+01  1.659000e+01   
std    7.079016e+01  5.001778e-01  2.558149e+02  1.236226e+00  8.291560e-02   

         tip_amount  tolls_amount  improvement_surcharge  total_amount  \
count  2.463931e+06  2.463931e+06           2.463931e+06  2.463931e+06   
mean   2.385696e+00  3.749773e-01           2.967234e-01  1.916937e+01   
min   -1.252200e+02 -3.140000e+01          -3.000000e-01 -4.803000e+02   
25%    7.200000e-01  0.000000e+00           3.000000e-01  1.130000e+01   
50%    2.000000e+00  0.000000e+00           3.000000e-01  1.469000e+01   
75%    3.000000e+00  0.000000e+00           3.000000e-01  2.002000e+01   
max    8.888800e+02  1.933000e+02           3.000000e-01  4.010956e+05   
std    2.830698e+00  1.680538e+00           4.374741e-02  2.559641e+02   

       congestion_surcharge   airport_fee  
count          2.392428e+06  2.392428e+06  
mean           2.282322e+00  8.249935e-02  
min           -2.500000e+00 -1.250000e+00  
25%            2.500000e+00  0.000000e+00  
50%            2.500000e+00  0.000000e+00  
75%            2.500000e+00  0.000000e+00  
max            2.500000e+00  1.250000e+00  
std            7.432040e-01  3.125554e-01

Yeah, there’s something odd in the dataset. For example, the columns for "fare_amount", "mta_tax", "tip_amount", "tolls_amount", and "airport_fee" contain negative values which is unexpected. These are all charges that should be positive or zero, not negative. It could be a typo, a system bug, or maybe a refund that wasn’t recorded properly. A good next step might be to either leave those rows out of the analysis or, if possible, try to get a cleaner version of the data.

**6. Drop all rows with NULL values and store the result. We'll ignore NULL valued rows in this mini-project.**

In [74]:
df_filter = df.dropna()

In [80]:
#We can see this works by either looking at the data directly or the shape reduction
print(df_filter.head())
print(f'{df.shape} -drop nulls-> {df_filter.shape}')

   VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2022-01-01 00:35:40   2022-01-01 00:53:29              2.0   
1         1  2022-01-01 00:33:43   2022-01-01 00:42:07              1.0   
2         2  2022-01-01 00:53:21   2022-01-01 01:02:19              1.0   
3         2  2022-01-01 00:25:21   2022-01-01 00:35:23              1.0   
4         2  2022-01-01 00:36:48   2022-01-01 01:14:20              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           3.80         1.0                  N           142           236   
1           2.10         1.0                  N           236            42   
2           0.97         1.0                  N           166           166   
3           1.09         1.0                  N           114            68   
4           4.30         1.0                  N            68           163   

   payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \


**Part 2**

**1. Create a new feature that calculates the trip duration in minutes.**